server/src/Icepeak/Server/WebsocketServer/Payload.hs

```diff
@@ -0,0 +1,274 @@
+        , "message" .= (message :: Text)
+        , "extra" .= extraData
+        ]
+
+      addonKeyValues :: [ Aeson.Pair ]
+      addonKeyValues =
+        case mbPaths of
+          Just paths -> [ "paths" .= paths ]
+          Nothing  -> []
+
+
+-- * Close Conn Reason
+
+data CloseType
+  = TypeSizeOutOfBounds
+  | TypeBinaryPayload
+  | TypeJsonDecodeError
+  | TypePayloadNotObject
+  | TypeMissingOrUnexpectedType
+  deriving (Show, Eq)
+
+-- https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/close#parameters
+
+-- Status codes in the range 3000-3999 are reserved for use by
+-- libraries, frameworks, and applications. The interpretation of these codes
+-- is undefined by this protocol.
+
+closeCode :: CloseType -> Word16
+closeCode TypeSizeOutOfBounds         = 3001
+closeCode TypeBinaryPayload           = 3002
+closeCode TypeJsonDecodeError         = 3003
+closeCode TypePayloadNotObject        = 3004
+closeCode TypeMissingOrUnexpectedType = 3005
+
+
+-- A string providing a custom WebSocket connection close reason (a concise human-readable prose explanation for the closure).
+-- The value must be no longer than 123 bytes (encoded in UTF-8).
+
+closeMessage :: CloseType -> Text
+closeMessage TypeSizeOutOfBounds
+  = "Received a payload size that is out of bounds"
+closeMessage TypeBinaryPayload
+  = "Received a payload using Binary instead of Text"
+closeMessage TypeJsonDecodeError
+  = "Received a payload that resulted in a JSON decode error"
+closeMessage TypePayloadNotObject
+  = "Received a JSON payload which is not an object"
+closeMessage TypeMissingOrUnexpectedType
+  = "Received a payload with a missing or unexpected value for 'type'"
+
+-- * Parsing
+
+maxPayloadBytes :: Int
+maxPayloadBytes = 1_000_000
+
+checkBounds
+  :: Lazy.ByteString
+  -> Either RequestMalformedError Lazy.ByteString
+checkBounds lazyBS = isOutOfBoundsAcc lazyBS 0
+  where
+    isOutOfBoundsAcc Lazy.ByteString.Empty _ = Right lazyBS
+    isOutOfBoundsAcc (Lazy.ByteString.Chunk chunk rest) accSize =
+      let accSize' = accSize + ByteString.length chunk in
+        if accSize' > maxPayloadBytes
+        then Left PayloadSizeOutOfBounds
+        else isOutOfBoundsAcc rest accSize'
+
+parseDataMessage
+  :: WebSockets.DataMessage -> RequestPayload
+parseDataMessage (WebSockets.Binary _ ) = RequestPayloadMalformed UnexpectedBinaryPayload
+parseDataMessage (WebSockets.Text utf8EncodedLazyByteString _ ) =
+  case parsedPayload of
+    (Left malformed) -> RequestPayloadMalformed malformed
+    (Right (Left subscribe)) -> RequestPayloadSubscribe subscribe
+    (Right (Right unsubscribe)) -> RequestPayloadUnsubscribe unsubscribe
+  where
+    parsedPayload
+      :: Either RequestMalformedError (Either RequestSubscribe RequestUnsubscribe)
+    parsedPayload = do
+      let
+        parseEither :: Aeson.Parser a -> Either String a
+        parseEither parser = Aeson.parseEither (const parser) ()
+
+        mapError = flip Bifunctor.first
```

Comment:
You might get some mileage from `mapError` instead of going through `Bifunctor`,
as you're already working within the `Either` monad down below to throw parsing errors.

Reply:
I'm not quite catching what you mean. There's already a a `mapError` function that exists?

Reply:
I haven't typed this into my editor so I may be wrong, but If I squint, I think your `mapError` is the same as [mtl's](https://hackage.haskell.org/package/mtl-2.3.1/docs/Control-Monad-Error-Class.html#v:mapError).

Though looking at the cabal bounds, we use upto <2.3 :(